File: 0032 - Longest Valid Parentheses/solution.c
#define max(a,b) (((a)>(b))?(a):(b))
int longestValidParentheses(char *s){
    int res = 0, l = 0, r = 0, i;
    for(i = 0; s[i]; i++){
        s[i]=='('?l++:r++;
        if(l == r) res = max(res, l+r);
        if(r > l) l = r = 0;
    }
    l = r = 0;
    for(i = i-1; i >= 0; i--){
        s[i]=='('?l++:r++;
        if(l == r) res = max(res, l+r);
        if(l > r) l = r = 0;
    }
    return res;
}

File: 0032 - Longest Valid Parentheses/solution.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        res, stk = 0, [-1]
        for i in range(len(s)):
            if s[i] == '(':
                stk.append(i)
            else:
                stk.pop()
                if not stk:
                    stk.append(i)
                else:
                    res = max(res, i-stk[-1])
        return res